core/server/master/src/main/java/alluxio/master/file/DefaultBlockDeletionContext.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.master.file;
+
+import alluxio.exception.status.UnavailableException;
+import alluxio.master.block.BlockMaster;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+/**
+ * Class for tracking which blocks need to be deleted, and deleting them on close.
+ */
+public final class DefaultBlockDeletionContext implements BlockDeletionContext {
+  private final BlockMaster mBlockMaster;
+  private final List<Long> mBlocks;
+
+  /**
+   * @param blockMaster the block master to use for deleting blocks
+   */
+  public DefaultBlockDeletionContext(BlockMaster blockMaster) {
+    mBlockMaster = blockMaster;
+    mBlocks = new ArrayList<>();
+  }
+
+  /**
+   * @param blockIds the blocks to be deleted when the context closes
+   */
+  @Override
+  public void registerBlocksForDeletion(Collection<Long> blockIds) {
+    mBlocks.addAll(blockIds);
+  }
+
+  /**
+   * @param blockId the block to be deleted when the context closes
+   */
+  @Override
+  public void registerBlockForDeletion(long blockId) {
+    mBlocks.add(blockId);
+  }
+
+  @Override
+  public void close() throws UnavailableException {
+    mBlockMaster.removeBlocks(mBlocks, true);
```

Comment:
Should we also clear `mBlocks` to be safe?

Reply:
done

Reply:
do we need to retry if one `UnavailableException` is thrown to ensure remaining blocks deleted?

Reply:
If an `UnavailableException` is thrown, we won't know which blocks have been deleted. However, since blocks can only be removed once, `removeBlocks` is actually idempotent so we can safely call it again, and it turns out clearing `mBlocks` doesn't matter too much.

Reply:
the only reason `removeBlocks` should fail is if we fail to write to the journal. There are already retries baked into the journal write, so I think it should be fine to let the exception propagate here

Reply:
actually I added 3 retries to be safe